### PR TITLE
[Interface] Return C-compatible result when Rust FFI function fails

### PR DESCRIPTION
**Pull request by satoshiotomakan**
## Description

Return C-compatible results (`CStrResult`, `CStrMutResult`, `CByteArrayResult`) from Rust FFIs.
For the full list of improvements see `Checklist`.

## How to test

Run C++/Rust tests

## Checklist

- [x] Return C-compatible result from `tw_encoding` FFIs
- [x] Return C-compatible result from `tw_move_parser` FFIs
- [x] Return C-compatible result from `tw_starknet` FFIs
- [x] Refactor CByteArray so it can be easily used inside Rust
- [x] Fix `tw_encoding::base32` implementation by using a well-tested [data-encoding](https://crates.io/crates/data-encoding) crate
- [x] Improve test coverage

### CODE REVIEW

**Review by Milerius — COMMENTED** [review]
minor comments / questions, otherwise good job

File: rust/tw_hash/src/ffi.rs [hunk 1]
```diff
@@ -16,7 +16,7 @@ use tw_memory::ffi::CByteArray;
 #[no_mangle]
 pub unsafe extern "C" fn blake_256(input: *const u8, input_len: usize) -> CByteArray {
     let input = std::slice::from_raw_parts(input, input_len);
-    blake::blake_256(input).into()
+    CByteArray::new(blake::blake_256(input))
```
[the remark this answers is not in the document]
Comment (satoshiotomakan) [reply rust/tw_hash/src/ffi.rs:19]: I've changed it in one of the previous commits. Will revert this, thanks!

File: rust/tw_memory/tests/c_byte_array_ffi_tests.rs [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+#[test]
+fn test_free_c_byte_array() {
+    unsafe {
+        free_c_byte_array(std::ptr::null_mut());
+
+        let mut raw_array = CByteArray::new(vec![1, 2, 3]);
+        free_c_byte_array(&mut raw_array as *mut CByteArray);
+
+        // The following leads to an undefined behaviour.
+        // free_c_byte_array(raw_array);
```
[the remark this answers is not in the document]
Comment (satoshiotomakan) [reply rust/tw_memory/tests/c_byte_array_ffi_tests.rs:18]: I've left this for transparency.

File: src/Aptos/TransactionPayload.cpp [hunk 3]
```diff
@@ -75,9 +75,14 @@ EntryFunction EntryFunction::from_json(const nlohmann::json& payload) noexcept {
     std::vector<Data> args;
     for (auto&& cur : payload.at("arguments")) {
         auto curStr = cur.get<std::string>();
-        auto* res = Rust::parse_function_argument_to_bcs(curStr.c_str());
-        args.emplace_back(parse_hex(res));
-        Rust::free_string(res);
+        auto res = Rust::parse_function_argument_to_bcs(curStr.c_str());
+        if (res.code != Rust::OK_CODE) {
+            // TODO consider exiting this function.
+            args.emplace_back();
```
[the remark this answers is not in the document]
Comment (satoshiotomakan) [reply src/Aptos/TransactionPayload.cpp:81]: There is always a possibility that we haven't covered all edge cases in unit tests. That may lead to an invalid transaction payload.
I would refactor this later and return at least an empty `EntryFunction` from `EntryFunction::from_json` if we got an error on argument parsing.

File: src/Base32.h [hunk 4]
```diff
@@ -20,22 +21,22 @@ inline bool decode(const std::string& encoded_in, Data& decoded_out, const char*
         return true;
     }
     auto decoded = Rust::decode_base32(encoded_in.c_str(), alphabet_in, false);
-    if (decoded.data == nullptr || decoded.size == 0) {
+    if (!Rust::data_from_c_byte_array_result(std::move(decoded), decoded_out)) {
```
[the remark this answers is not in the document]
Comment (satoshiotomakan) [reply src/Base32.h:24]: Good idea, will do!

File: src/rust/Wrapper.h [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+#include <optional>
+
+#include "Data.h"
+#include "rust/bindgen/WalletCoreRSBindgen.h"
+
+namespace TW::Rust {
+
+class CByteArrayWrapper {
+public:
```
Comment (Milerius) [inline src/rust/Wrapper.h:17]: can be a struct if everything is public

File: src/Base32.h [hunk 6]
```diff
@@ -20,11 +20,12 @@ inline bool decode(const std::string& encoded_in, Data& decoded_out, const char*
     if (encoded_in.empty()) {
         return true;
     }
-    auto decoded = Rust::decode_base32(encoded_in.c_str(), alphabet_in, false);
-    if (!Rust::data_from_c_byte_array_result(std::move(decoded), decoded_out)) {
-        return false;
+    Rust::CResult<Rust::CByteArrayWrapper> res = Rust::decode_base32(encoded_in.c_str(), alphabet_in, false);
```
Comment (Milerius) [inline src/Base32.h:23]: Can have a sugar using in Wrapper.h -> `using CByteArrayResult = Rust::CResult<Rust::CByteArrayWrapper>`

**Review by lamafab — APPROVED** [review]
